### PR TITLE
boards/blxxxpill: configure usbdev_fs

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This attempts to enable the `usbdev_fs` module on STM32F1.
Unfortunately this does not work yet - It builds, but it looks like something else needs to be enabled on this platform.
Nothing shows up in `dmesg` when the USB test code is run (also no errors).

I don't know where to start looking for what it could be, but looking for some pointers to get USB support for this popular board.

### Testing procedure

Flash e.g. `tests/usbus_hid` or `tests/usbus_cdc_acm_stdio` on a bluepill or blackpill board.

### CODE REVIEW

**Comment by maribu** [comment]
Are you aware of the wrong pull resistor strength bug of some early bluepill boards? Most USB hosts just worked anyway, but if you have such a board and a nitpicky machine, you might have to replace two resistors on the board.

**Comment by dylad** [comment]
My first guess (without a proper look on datasheet) would be it needs some clock configuration to enable this IP.
Maybe @bergzand knows better ?

**Comment by benpicco** [comment]
Well I do get some interrupts (just `0x400` - RESET with this, when I remove the periph reset in the driver I get `0x2400` and `0x2100`) so the peripheral must be clocked.

The USB interface works pretty much immediately with [this minimal example](https://github.com/Erlkoenig90/f1usb/tree/minimal).

File: boards/common/blxxxpill/include/periph_conf.h [hunk 1]
```diff
@@ -310,6 +310,32 @@ static const spi_conf_t spi_config[] = {
...
+ * @brief Interrupt function name mapping
+ */
+#define USBDEV_ISR             isr_usb_lp_can1_rx0
+
+/**
+ * @brief Number of available USB device FS peripherals
+ */
+#define USBDEV_NUMOF           ARRAY_SIZE(stm32_usbdev_fs_config)
+
+
```
Comment (maribu) [inline boards/common/blxxxpill/include/periph_conf.h:338]:
```suggestion

```

File: boards/common/blxxxpill/Makefile.features [hunk 2]
```diff
@@ -11,3 +11,4 @@ FEATURES_PROVIDED += periph_spi
 FEATURES_PROVIDED += periph_timer
 FEATURES_PROVIDED += periph_uart
 FEATURES_PROVIDED += periph_qdec
+FEATURES_PROVIDED += periph_usbdev
```
Comment (maribu) [inline boards/common/blxxxpill/Makefile.features:14]: If you don't have time to look into, I would be fine to merge this as is with just the feature disabled and a comment. Maybe someone else would pick this up from there.

```suggestion
# TODO: Figure out why the USB driver doesn't work
# FEATURES_PROVIDED += periph_usbdev
```

**Review by maribu — COMMENTED** [review]
see inline

File: boards/common/blxxxpill/include/periph_conf.h [hunk 3]
```diff
@@ -310,6 +310,32 @@ static const spi_conf_t spi_config[] = {
...
 /** @} */
 
+/**
+ * @brief USB device FS configuration
+ */
+static const stm32_usbdev_fs_config_t stm32_usbdev_fs_config[] = {
+    {
+        .base_addr  = (uintptr_t)USB,
+        .rcc_mask   = RCC_APB1ENR_USBEN,
+        .irqn       = USB_LP_IRQn,
```
Comment (gschorcht) [inline boards/common/blxxxpill/include/periph_conf.h:320]:
```suggestion
        .irqn       = USB_LP_CAN1_RX0_IRQn,
```

File: boards/common/blxxxpill/include/periph_conf.h [hunk 4]
```diff
@@ -310,6 +310,32 @@ static const spi_conf_t spi_config[] = {
...
+        .apb        = APB1,
+        .dm         = GPIO_PIN(PORT_A, 11),
+        .dp         = GPIO_PIN(PORT_A, 12),
+        .af         = GPIO_AF_OUT_OD,
```
Comment (gschorcht) [inline boards/common/blxxxpill/include/periph_conf.h:324]:
```suggestion
        .af         = GPIO_AF_UNDEF,
```
According to the reference manual, these pins are connected automatically to the USB internal transceiver connected to the as soon as the USB is enabled. This change should work with PR #18680.

**Comment by benpicco** [comment]
Hm, still not much luck with this one.

**Comment by gschorcht** [comment]
> Hm, still not much luck with this one.

I'm not familiar with the clock hierarchy in STM32 MCUs, but If I'm right, the `bluepill` board uses the 8 MHz HSE clock, 72 MHz as PLL clock and as AHB clock, and 38 MHz as APB1 clock. Since the USB clock is derived from the AHB clock, it would have to be pre-divided by 1.5 by
```c
RCC->CFGR |= RCC_CFGR_USBPRE;
```
which is not yet part of the driver.

Could you check whether bit 22 is set before the USB peripheral clock is enabled?

**Comment by gschorcht** [comment]
> Could you check whether bit 22 is set before the USB peripheral clock is enabled?

On the other hand, I am a bit confused because the `nucleo-f303ze` board uses the same clock configuration and `RCC_CFGR_USBPRE` is already set, although the reset value of `RCC->CFGR` is `0x0`.

**Comment by benpicco** [comment]
`RCC->CFGR` is `5d040a`, so the `RCC_CFGR_USBPRE` bit (`0x400000`) is set.

I also get a single reset (`0x400`) interrupt from the device, but nothing after that.

**Comment by gschorcht** [comment]
> `RCC->CFGR` is `5d040a`, so the `RCC_CFGR_USBPRE` bit (`0x400000`) is set.
> 
> I also get a single reset (`0x400`) interrupt from the device, but nothing after that.

Hm, that is, when a RESET interrupt comes, the controller has has pulled up a data line and the host has pushed it back down to send the USB RESET signal. Do you have any chance to record the USB DM and DP signals PA11 and PA12 with a logic analyzer to see
- whether the USB clock is correct and
- whether the Blue Pill receives the subsequent SETUP/DAT0 packets from the hosts and acknowledges them with an ACK?

Unfortunatly, I have only a Blue Pill board with the Chinese STM32 clone CH32F103 from WCH which is not flashable with ST-Link :worried:

**Comment by gschorcht** [comment]
> Do you have any chance to record the USB DM and DP signals PA11 and PA12 with a logic analyzer to see

I have tried to reproduce the `bluepill` USB problem with a `nucleo-f103rb`. Although the USB controller is enabled and the clock seems to be configured correctly, none of the USB signals are pulled up, so the host does not recognize the device.

If an USB signal is pulled up by a device and the host recognizes this, there should be printed at least something like
```
Oct  2 14:14:40 gunny8 kernel: [1906254.288780] usb 1-4.4: new high-speed USB device number 96 using xhci_hcd
```
in the syslog even if the device misbehaves.

**Comment by gschorcht** [comment]
For testing, I tried a little trick by temporarily configuring the USB_DM GPIO as an output and pushing it down for about 20 us during initialization and then configuring it as an input again so that the resistor pulls up the USB_DM signal back. This simulates detaching and attaching the device so that the host at least recognizes the device and starts sending SETUP packets. These SETUP packets are not ACKed by the device. One reason might be that EP0 isn't enabled. In tinyUSB stm32_fsdev driver, EP0 is enabled during initialization.

**Comment by dylad** [comment]
> One reason might be that EP0 isn't enabled. In tinyUSB stm32_fsdev driver, EP0 is enabled during initialization.

You are suggesting that EP0 init is done too late on bluepill ? That might be possible. Looking at the schematic, the pullup is hardwired, no way to control it. So we might need a special handling to trigger a detach/attach sequence just before EP0 initialization in usbdev_fs driver.

I really should order a bluepill...

**Comment by gschorcht** [comment]
> You are suggesting that EP0 init is done too late on bluepill ?

No, not yet :wink: It's still too early to conclude anything. I was just reporting my observations and what the tinyUSB driver does. I will investigate it further with my "Nucleo-f103rb" board. I will get two `bluepill` boards today.

**Comment by gschorcht** [comment]
After some further debugging, it could be that I found one reason why the driver doesn't work for STM32F103. There seems to be a problem with the half-word memory accesses to the USB IP Packet Memory Area here
https://github.com/RIOT-OS/RIOT/blob/f30f9453f725cbc4f31e911ff48ecb9b2399d3ae/cpu/stm32/periph/usbdev_fs.c#L507-L509 where  `EP_DESC[ep->num]` is defined as structure of type https://github.com/RIOT-OS/RIOT/blob/f30f9453f725cbc4f31e911ff48ecb9b2399d3ae/cpu/stm32/periph/usbdev_fs.c#L69-L75 `EP_DESC` simply starts at the USB IP Packet Memory Area base address (`USB_PMADDR = 0x40006000`). The source code above produces the following half-word memory access instructions
```asm
strh   r1, [r3, #4]
strh   r2, [r3, #6]
```
where the register values before the executions are:
```
r1             0x80
r2             0x8400
r3             0x40006000
```
`r1` and `r2` contain the right values and `r3` points to `USB_PMADDR`. These instrcutions should lead to the following results in the USB Packet Memory Area.
```
0x40006000:     0x40    0x00    0x00    0x00    0x80    0x00    0x00    0x84
```
The result is however:
```
0x40006000:     0x40    0x00    0x00    0x00    0x00    0x84    0x00    0x00
```
If the code is executed stepwise, it can be observed that the correct value is in the memory after the first half-word store instruction
```
strh   r1, [r3, #4]
```
```
0x40006000:     0x40    0x00    0x00    0x00    0x80    0x00    0x00    0x00
```
but is overwritten by the second half-word instruction:
```
strh   r2, [r3, #6]
```
```
0x40006000:     0x40    0x00    0x00    0x00    0x00    0x84    0x00    0x00
```
So I guess, the cause for the problem might be that the USB IP Packet Memory Area can be accessed only word-aligned.

**Comment by dylad** [comment]
Wow... What sorcery is this ?!
When I started this driver (for WB family). I wrote `ep_buf_desc_t` struct while looking at the [reference manual](https://www.st.com/resource/en/reference_manual/rm0434-multiprotocol-wireless-32bit-mcu-armbased-cortexm4-with-fpu-bluetooth-lowenergy-and-802154-radio-solution-stmicroelectronics.pdf) part 39.6.2 Buffer descriptor table
There, they describe the members of the struct as:
USB_ADDRn_TX     ([USB_BTABLE] + n*8)
USB_COUNTn_TX  ([USB_BTABLE] + n*8 + 2)
USB_ADDRn_TX     ([USB_BTABLE] + n*8 + 4)
USB_COUNTn_RX  ([USB_BTABLE] + n*8 + 6)

Thus, I assumed the struct should be filled with uint16_t values.

Now, looking at [STM32F103 Reference Manual](https://www.st.com/resource/en/reference_manual/rm0008-stm32f101xx-stm32f102xx-stm32f103xx-stm32f105xx-and-stm32f107xx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf), part 25.5.3 Buffer descriptor table
It tells a quite different story...
```
Due to the common limitation of APB1
bridges on word addressability, all register addresses are aligned to 32-bit word boundaries
although they are 16-bit wide. The same address alignment is used to access packet buffer
memory locations, which are located starting from 0x4000 6000
```
So the offets became:
Address offset: [USB_BTABLE] + n*16
USB local address: [USB_BTABLE] + n*8

:exploding_head:

**Comment by gschorcht** [comment]
[STM32F303 Reference Manual](https://www.st.com/resource/en/reference_manual/rm0316-stm32f303xbcde-stm32f303x68-stm32f328x8-stm32f358xc-stm32f398xe-advanced-armbased-mcus-stmicroelectronics.pdf), section 32.6.2 even describes both access methods.

Do we have any define in CMSIS headers by which we could determine the access scheme?

**Comment by gschorcht** [comment]
> [STM32F303 Reference Manual](https://www.st.com/resource/en/reference_manual/rm0316-stm32f303xbcde-stm32f303x68-stm32f328x8-stm32f358xc-stm32f398xe-advanced-armbased-mcus-stmicroelectronics.pdf), section 32.6.2 even describes both access methods.

According to [STM32F303 Reference Manual](https://www.st.com/resource/en/reference_manual/rm0316-stm32f303xbcde-stm32f303x68-stm32f328x8-stm32f358xc-stm32f398xe-advanced-armbased-mcus-stmicroelectronics.pdf), section 32.3 Table 175, STM32F303xB/C MCUs have 512 bytes packet buffer memory and word-aligned access while STM32F303xD/E MCUs have 1024 bytes packet buffer memory and half-word access.

**Comment by gschorcht** [comment]
Ok, after all I have a modified driver version using the "1 x 16 bits/word" access scheme which is working now on STM32F103. Do we have any chance to determine which access scheme should be compiled in without using a bunch of `#ifdef`s for all families and lines?

**Comment by gschorcht** [comment]
> Do we have any define in CMSIS headers by which we could determine the access scheme?

If not, we could define the access scheme and/or PMA size as part of the board configuration in `periph_conf.h`.

**Comment by dylad** [comment]
> Do we have any define in CMSIS headers by which we could determine the access scheme?

I would say no.
I just looked at how this madness is managed by ST itself and tinyUSB.
They both create a variable to align their access 
For ST, they have a `PMA_ACCESS` variable declared per family in STM32Cube header.
[which is 1 for WB](https://github.com/STMicroelectronics/STM32CubeWB/blob/83aacadecbd5136ad3194f39e002ff50a5439ad9/Drivers/STM32WBxx_HAL_Driver/Inc/stm32wbxx_ll_usb.h#L169)
[which is 2 for F1](https://github.com/STMicroelectronics/STM32CubeF1/blob/218b5100dd9e4b9bb3b0aa7fc77467373e932025/Drivers/STM32F1xx_HAL_Driver/Inc/stm32f1xx_ll_usb.h#L523)
On TinyUSB side, they also have a variable called `PMA_STRIDE` which is defined [there](https://github.com/hathach/tinyusb/blob/a95fb36f72730dc22cb73be379f8e2dacb2b4e75/src/portable/st/stm32_fsdev/dcd_stm32_fsdev_pvt_st.h#L110).

**Comment by gschorcht** [comment]
> They both create a variable to align their access

We should do it in same way. The questions is whether we should define it on board level or on MCU level. tinyUSB defines the `PMA_LENGTH` on MCU level which in turn is used to define the `PMA_STRIDE`.

**Comment by benpicco** [comment]
It's a CPU property, the board has no influence on this.

**Comment by gschorcht** [comment]
> It's a MCU property, the board has no influence on this.

I will provide my changes of the driver next days (the change of the access scheme and a further change that is necessary to get the Bluepill working), I'm busy today.

**Comment by dylad** [comment]
@benpicco Do you think we can have this one merged before feature freeze ?

File: cpu/stm32/Makefile.dep [hunk 5]
```diff
@@ -4,7 +4,7 @@
 USEMODULE += periph stm32_clk stm32_vectors
 
 ifneq (,$(filter periph_usbdev,$(FEATURES_USED)))
-  ifeq (,$(filter f3 wb,$(CPU_FAM)))
+  ifeq (,$(filter f1 f3 wb,$(CPU_FAM)))
```
Comment (gschorcht) [inline cpu/stm32/Makefile.dep:7]: I think we have to split it for F1 family, STM32F105xx and STM32F107xx use USB OTG IP core. We could merge it for now as is.

**Review by gschorcht — APPROVED** [review]
Changes are straight forward and the `bluepill` board works.

**Comment by benpicco** [comment]
let's drop da2473c5f8 for now

**Comment by gschorcht** [comment]
> let's drop [da2473c](https://github.com/RIOT-OS/RIOT/commit/da2473c5f822d5649118b8274a6dc07c97693710) for now

Doesn't it work or what is the reason for that? Using the USB connector for the console would make the life much easier.

**Comment by benpicco** [comment]
No it works fine. The problem is that it pushes some apps past the limit on the smaller boards and there was also a Kconfig issue remaining.

I simply dropped the default STDIO change to get this in before the feature freeze, then provide it in a follow up PR.

**Comment by maribu** [comment]
Please extend the Makefile.ci as required and squash. Compile test can be skipped this time.

**Pull request merged**